orm_lib/inc/drogon/orm/BaseBuilder.h

```diff
@@ -0,0 +1,337 @@
+            sql += " offset " + std::to_string(offset_.value());
+        }
+        return sql;
+    }
+
+    inline std::vector<std::string> gen_args() const
+    {
+        std::vector<std::string> args;
+        if (!filters_.empty())
+        {
+            for (const Filter& f : filters_)
+            {
+                args.emplace_back(f.value);
+            }
+        }
+        return args;
+    }
+
+  public:
+#ifdef __cpp_if_constexpr
+    static ResultType convert_result(const Result& r)
+    {
+        if constexpr (SelectAll)
+        {
+            if constexpr (Single)
+            {
+                return T(r[0]);
+            }
+            else
+            {
+                std::vector<T> ret;
+                for (const Row& row : r)
+                {
+                    ret.emplace_back(T(row));
+                }
+                return ret;
+            }
+        }
+        else
+        {
+            if constexpr (Single)
+            {
+                return r[0];
+            }
+            else
+            {
+                return r;
+            }
+        }
+    }
+#else
+    template <bool SA = SelectAll,
+              bool SI = Single,
+              std::enable_if_t<SA, std::nullptr_t> = nullptr,
+              std::enable_if_t<SI, std::nullptr_t> = nullptr>
+    static inline T convert_result(const Result& r)
+    {
+        return T(r[0]);
+    }
+    template <bool SA = SelectAll,
+              bool SI = Single,
+              std::enable_if_t<SA, std::nullptr_t> = nullptr,
+              std::enable_if_t<!SI, std::nullptr_t> = nullptr>
+    static inline std::vector<T> convert_result(const Result& r)
+    {
+        std::vector<T> ret;
+        for (const Row& row : r)
+        {
+            ret.template emplace_back(T(row));
+        }
+        return ret;
+    }
+    template <bool SA = SelectAll,
+              bool SI = Single,
+              std::enable_if_t<!SA, std::nullptr_t> = nullptr,
+              std::enable_if_t<SI, std::nullptr_t> = nullptr>
+    static inline Row convert_result(const Result& r)
+    {
+        return r[0];
+    }
+    template <bool SA = SelectAll,
+              bool SI = Single,
+              std::enable_if_t<!SA, std::nullptr_t> = nullptr,
+              std::enable_if_t<!SI, std::nullptr_t> = nullptr>
+    static inline Result convert_result(const Result& r)
+    {
+        return r;
+    }
+#endif
+
+    inline ResultType execSync(const DbClientPtr& client)
+    {
+        Result r(nullptr);
+        {
+            auto binder = *client << gen_sql(client->type());
+            for (const std::string& a : gen_args())
+            {
+                binder << a;
+            }
+            binder << Mode::Blocking;
+            binder >> [&r](const Result& result) { r = result; };
+            binder.exec();  // exec may throw exception
+        }
+        return convert_result(r);
+    }
+
+    std::future<ResultType> execAsync(const DbClientPtr& client)
```

Comment:
`execAsync` should takes callbacks as arguments. Current `execAsync` api should be renamed to something like `execAsyncFuture`

Reply:
@hwc0919 Thank you! Would you need an `execAsync` implementation as well?